elytron/WFCORE-5178_native-support-for-oidc.adoc

```diff
@@ -0,0 +1,178 @@
+The new subsystem will install a `DeploymentUnitProcessor` which will detect if OIDC is required for a deployment.
+
+Like with the existing Keycloak client adapter, the OIDC configuration will be specified via the deployment
+or via subsystem configuration. For the deployment case, the OIDC configuration will be specified in a JSON
+configuration file in the WEB-INF directory along with the `auth-method`, `OIDC`, specified in the `web.xml`.
+For the subsystem case, the `auth-method` won't need to be specified in the `web.xml` file, it will be set
+automatically.
+
+Once a deployment has been processed, if its `auth-method` is `OIDC`, the OIDC integration will be activated.
+Otherwise, no activation will occur and the deployment will continue to use the existing security capabilities
+provided by WildFly.
+
+If the OIDC integration is activated for a deployment, then appropriate module dependencies will be added to
+the deployment (e.g., a dependency on the new Elytron OIDC module). This will ensure the activation of the
+OIDC HTTP mechanism implementation for the deployment. The deployment will make use of a "Virtual Security Domain",
+similar to what was implemented for https://github.com/wildfly/wildfly-proposals/blob/master/microprofile/WFLY-12609_MicroProfile_JWT_1_1.adoc#hard-requirements[Microprofile JWT].
+This means that it will no longer be necessary to pre-configure SecurityDomains in the Elytron subsystem,
+as was required with the Keycloak client adapter.
+
+It should be possible to make use of different OpenID providers with WildFly. While working on the
+implementation for this RFE, we should make sure that we aren't including Keycloak-specific configuration, if
+possible.
+
+==== New galleon layer for OIDC support
+
+A new Galleon layer, `oidc`, should be added to allow a server to be provisioned with OIDC support.
+
+=== Nice-to-Have Requirements
+
+* https://openid.net/specs/openid-connect-rpinitiated-1_0.html[RP-Initiated Logout] is an optional
+OIDC specification that defines how a relying party (RP) can request that an OpenID provider logs
+out the end-user. The Keycloak client adapter provides support for this so this is something we
+should provide as well but we could possibly look at that as a follow-up to this RFE if necessary.
+
+* The existing Keycloak adapter can also be used to secure the admin console. This requires some
+additional work so it might be done as a follow-up task.
+
+* *Migration*: For the most part, the same subsystem attributes and JSON configuration options that could
+be used with the `keycloak-adapter-subsystem` and the `keycloak.json` file will still work with the
+`elytron-oidc-subsystem` and the `oidc.json` file. The main difference is the actual subsystem name and
+the name of the JSON configuration file. That means that the migration tool could potentially detect an
+existing `keycloak` subsystem configuration and then change its name to the new subsystem
```

Comment:
What about attributes like `client-key-password`, `client-keystore`, `client-keystore-password`, `realm-public-key`, `truststore`, `truststore-password`? We should transform those to leverage (Elytron) capabilities, like client `ssl-context` and/or `credential-store`.

Reply:
Right now, it's possible to specify these attributes in either the oidc.json file or in the subsystem configuration.

The advantage of the first approach is that all of the configuration is specified in the deployment itself without requiring any server configuration.

I could look into the possibility of updating the `elytron-oidc-client` subsystem configuration to allow references to an Elytron `ssl-context` / `credential-store` to be specified. For the deployment case, perhaps the SSL config could be specified via a wildfly-config.xml file in the deployment.

I'll need to look into this further. This would likely need to be a follow-up task though since the original plan was to make use of the same options the Keycloak adapter provides.